toolchain/lex/tokenized_buffer.cpp

```diff
@@ -247,18 +247,48 @@ class TokenizedBuffer::Lexer {
         translator_(&buffer),
         emitter_(translator_, consumer),
         token_translator_(&buffer),
-        token_emitter_(token_translator_, consumer),
-        current_line_(buffer.AddLine(LineInfo(0))),
-        current_line_info_(&buffer.GetLineInfo(current_line_)) {}
+        token_emitter_(token_translator_, consumer) {}
+
+  // Find all line endings and create the line data structures. Explicitly kept
+  // out-of-line because this is a significant loop that is useful to have in
+  // the profile and it doesn't simplify by inlining at all. But because it can,
+  // the compiler will flatten this otherwise.
+  [[gnu::noinline]] auto CreateLines(llvm::StringRef source_text) -> void {
+    // We currently use `memchr` here which typically is well optimized to use
+    // SIMD or other significantly faster than byte-wise scanning. We also use
+    // carefully selected variables and the `ssize_t` type for performance and
+    // code size of this hot loop.
+    //
+    // TODO: Eventually, we'll likely need to roll our own SIMD-optimized
+    // routine here in order to handle CR+LF line endings, as we'll want those
+    // to stay on the fast path. We'll also need to detect and diagnose Unicode
+    // vertical whitespace. Starting with `memchr` should give us a strong
+    // baseline performance target when adding those features.
+    const char* const text = source_text.data();
+    const ssize_t size = source_text.size();
+    ssize_t start = 0;
+    while (const char* nl = reinterpret_cast<const char*>(
+               memchr(&text[start], '\n', size - start))) {
+      ssize_t nl_index = nl - text;
+      buffer_->AddLine(LineInfo(start, nl_index - start));
+      start = nl_index + 1;
+    }
+    // The last line ends at the end of the file.
+    buffer_->AddLine(LineInfo(start, size - start));
+
+    // Now that all the infos are allocated, get a fresh pointer to the first
+    // info for use while lexing.
+    current_line_ = Line(0);
+    current_line_info_ = &buffer_->GetLineInfo(current_line_);
+  }
 
   // Perform the necessary bookkeeping to step past a newline at the current
   // line and column.
   auto HandleNewline() -> void {
-    current_line_info_->length = current_column_;
-
-    current_line_ = buffer_->AddLine(
-        LineInfo(current_line_info_->start + current_column_ + 1));
+    int next_start = current_line_info_->start + current_column_ + 1;
+    current_line_ = buffer_->GetNextLine(current_line_);
     current_line_info_ = &buffer_->GetLineInfo(current_line_);
```

Comment:
I'm guessing there probably isn't a measurable difference from this tweak, but just in case:
```suggestion
    ++current_line_info_;
    CARBON_DCHECK(current_line_info_ == &buffer_->GetLineInfo(current_line_));
```

Reply:
No, the slow part is incrementing current_line_ -- I have some ideas to improve it eventually, but no need to be cleverer about the info in this PR. Let's revisit before we add dchecks and such.